lib/iex/lib/iex/introspection.ex

```diff
@@ -159,21 +153,35 @@ defmodule IEx.Introspection do
     end
   end
 
-  defp rewrite_elixir_source(module, source) do
+  @elixir_apps ~w(eex elixir ex_unit iex logger mix)a
+  @otp_apps ~w(kernel stdlib)a
+  @apps @elixir_apps ++ @otp_apps
+
+  defp rewrite_source(module, source) do
     case :application.get_application(module) do
-      {:ok, app} when app in [:eex, :elixir, :ex_unit, :iex, :logger, :mix] ->
-        {in_app, [lib_or_src | _]} =
-          source
-          |> Path.split()
-          |> Enum.reverse()
-          |> Enum.split_while(& &1 not in ["lib", "src"])
-
-        Application.app_dir(app, Path.join([lib_or_src | Enum.reverse(in_app)]))
+      {:ok, app} when app in @apps ->
+        Application.app_dir(app, rewrite_source(source))
       _ ->
-        List.to_string(source)
+        beam_path = :code.which(module)
```

Comment:
as discussed before, this can return an atom but I couldn't get it into that state (if module is not present or something, it's catched earlier)

Reply:
Elixir modules that are not compiled and only loaded in memory have this issue:

```elixir
defmodule Foo do
end
:code.which(Foo)
```

We need to check for the return otherwise it is a problem waiting to happen.




 I still think though we should check for it.

Reply:
good call!

1. in-memory module is handled in https://github.com/elixir-lang/elixir/pull/6616/commits/3538c30735e48c6d454353de9627b017535da8ab.
2. couldn't find a function to grab source file for preloaded module so just added explicit test in dfe694bf78e00acd6aadd4f6d4686b45b3fbc252
3. the only other value that `which` can return is `:cover_compiled` but not sure if we need to handle it.